ACE backend: a code generator from AirScript to the ACE chiplet of Miden's recursive verifier

paracetamolo opened the pull request:
This backend generates a table in the format required by the ACE chiplet as described in https://github.com/0xPolygonMiden/air-script/issues/366.

Currently the table is in the Graphviz dot format, the textual form is a superset of what is needed for the ACE chiplet and can be used to verify that rows are in the right order. The image generated can be used to debug the connections in the graph.

```
cd codegen/ace
cargo test -- test_ace && dot -Tsvg ace.dot > ace.svg
```

Al-Kindi-0 commented:
Do we need to change the base branch from `main`?

paracetamolo commented:
> Do we need to change the base branch from `main`?

It's based on a branch on the Massa Labs repo `massa/feat-implement-ir`, github doesn't show it for this repo.

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -62,6 +62,50 @@ fn table_to_string(table: Vec<Operation>) -> Result<String, std::fmt::Error> {
+        | Operation::Value(Value::RandomValue(_))
+        | Operation::Value(Value::TraceAccess(_)) => 0u64 << 60,
+        Operation::Value(Value::Constant(_)) => 1u64 << 60,
+        Operation::Value(Value::PeriodicColumn(_)) => panic!("periodic"),
```
I added a comment in the issue on a proposal to tackle `Value::PeriodicColumn`

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -62,6 +62,50 @@ fn table_to_string(table: Vec<Operation>) -> Result<String, std::fmt::Error> {
+        Operation::Value(Value::PublicInput(_))
+        | Operation::Value(Value::RandomValue(_))
+        | Operation::Value(Value::TraceAccess(_)) => 0u64 << 60,
+        Operation::Value(Value::Constant(_)) => 1u64 << 60,
```
We need to also serialize the constant value

paracetamolo replied:
done

Al-Kindi-0 reviewed (commented):
As a draft, looks great, thank you!
Could we add some tests to test for the correct evaluation on some small examples?
For example, and for small circuits, we can compare the output of the circuit to the value we get from the algebraic expression.

paracetamolo commented:
Added support for periodic columns as described [here](https://github.com/0xPolygonMiden/air-script/issues/366#issuecomment-2697324777), this required a bit of a refactor that overall improved the quality of the code.

bobbinth commented:
@paracetamolo - seems like this PR includes changes from other PRs. It would be good to rebase it from the latest `next` branch. This way, it'll contain only the relevant changes.

paracetamolo commented:
Done. I had rebased the branch on `next` a while ago but forgot to change the base of the PR.

paracetamolo commented:
Things left to do:
- test periodic columns. Most of the AirScript examples used in the MASM backend tests are also tested in the ACE backend, except for the ones using periodic columns.
- incorporate the recent changes described in https://github.com/0xPolygonMiden/air-script/issues/366#issuecomment-2751848648

bobbinth commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,257 @@
+/// - z
+/// - z^n
+/// - g^(-1)
+pub fn build_ace_circuit(ir: &Air) -> anyhow::Result<(NodeIndex, Circuit)> {
```
nit: this seems to be the main function in this module - I'd move it to the top of the file. The other functions/structs could go into helper sections after it.

bobbinth commented on codegen/ace/src/serialization.rs:
```diff
@@ -0,0 +1,138 @@
+/// Serialization to field elements
+/// Note that some information is lost, so the original circuit cannot be recovered from its felt serialization
+
+// The 63 bits available in a field element are used in the following order,
+// from most significant to least:
+// - 3 bits for the OPCODE, with READ 0, CONST 1, ADD 2, MUL 3, SUB 4
+// - 20 bits for out
+// - 20 bits for left
+// - 20 bits for right
+// Left and right are only used for ADD MUL and SUB and are expected to be zero
+// in the other cases.
+// When a felt encodes an opcode CONST and its out position, the following felt
+// must encode the 64bit constant.
+pub fn operation_to_felts(out: usize, op: Operation) -> Vec<Felt> {
```
This needs to be updated to be consistent with the lates specs.

Also, shouldn't this return a single field element? I think the only reason we have a vector as the return value is to handle the `Constant` operation, right? If so, is the current treatment of constants correct? I though they would be provided in a way similar to inputs - i.e., there would not be a lot of difference between inputs and constants.

paracetamolo replied:
The serialization you reviewed was the old one. Now I updated it and, as you mentioned, constants correspond to a single field element.

bobbinth commented on codegen/ace/src/serialization.rs:
```diff
@@ -0,0 +1,138 @@
+    }
+}
+
+fn operation_to_dot(out: NodeIndex, op: &Operation) -> Result<String, std::fmt::Error> {
```
Should this be a method of the `Operation` struct? - e.g.,

```rust
impl Operation {

    pub fn to_dot(&self, out: NodeIndex) -> Result<String, std::fmt::Error> {
        ...
    }

}
```
Let's also add doc comments to this function.

paracetamolo replied:
It is just an private internal function for `operations_to_dot`, not really worth exposing it.
I documented the functions.

bobbinth commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,436 @@
+    }
+}
+
+#[test]
+fn test_sum() {
```
nit: we usually put tests into `tests` module.

paracetamolo replied:
done

bobbinth commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,436 @@
+use std::cmp::Ordering;
+use std::collections::{btree_map::BTreeMap, btree_set::BTreeSet};
+
+pub struct Circuit(Vec<Operation>);
```
It may make sense to make the circuit a bit more structured:

- Have a field that describes the shape of expected inputs and constants.
- Maybe have something like the `Operation` struct but specifically tailored to the needs of this crate which more closely maps to the gate specifications.
- Arrange the operations in the order in which they will appear in the serialized form.

paracetamolo replied:
In this PR I was trying to reuse as much as possible `air_ir` where the `Operation` type comes from. It is the type that the ACE backend receives and, after a lot of processing, also the one that it outputs. For the output part we don't need all that structure so we can define a simpler type closer to the serialization format but I would leave this as a future PR so we can agree on the right design with @adr1anh.

bobbinth reviewed (commented):
Thank you! Looks good! I left some comments inline. Many of them are about improving code organization to be more inline with our usual style, but I also think that this needs to be updated to be inline with the latest ACE chiplet design.

paracetamolo commented on codegen/ace/0.dot:
They are regression files to detect when we change the compilation output. I was looking for a "Rust way" to place these files inside the test directory, I ended up using the environment variable `CARGO_MANIFEST_DIR`, not sure if there is a better way. Now they are placed in `codegen/ace/tests/regressions/`.

paracetamolo commented on codegen/ace/tests/utils.rs:
```diff
@@ -0,0 +1,26 @@
+};
+use std::sync::Arc;
+
+pub fn codegen(source: &str) -> (NodeIndex, Circuit) {
```
done

paracetamolo commented:
Update:
- I addressed most comments.
- Converted `eval` and tests to `QuadExtension`.
- `z^min_num_cycles` now is always an input, even if the circuit does not have periodic columns.
- I fixed a bug where unused inputs would be optimized away. Now inputs are always preserved, even if unused.

Todo:
- update formula as in point 7 of https://github.com/0xPolygonMiden/air-script/issues/366#issuecomment-2751848648
- fix serialization format

adr1anh commented on codegen/ace/src/serialization.rs:
```diff
@@ -0,0 +1,181 @@
+use miden_core::{Felt, FieldElement, QuadExtension};
+use std::fmt::Write;
+
+pub type Quad = QuadExtension<Felt>;
+pub fn to_quad(e: u64) -> Quad {
+    QuadExtension::new(Felt::new(e), Felt::ZERO)
+}
```
Nit: The type `QuadExtension` implements `From<u64>` which could be invoked instead of this methd

paracetamolo replied:
done

adr1anh commented on codegen/ace/src/comparison.rs:
```diff
@@ -0,0 +1,64 @@
+
+/// order by segment, offset, then column
+fn compare_access(a: &TraceAccess, b: &TraceAccess) -> Ordering {
+    let segment = a.segment.cmp(&b.segment);
+    if segment == Ordering::Equal {
+        let row = a.row_offset.cmp(&b.row_offset);
+        if row == Ordering::Equal {
+            a.column.cmp(&b.column)
+        } else {
+            row
+        }
+    } else {
+        segment
+    }
```
Nit: Tuples implement `Ord` if all its components implement `Ord`. If you need a different ordering of the structs, this should be simpler
```suggestion
    (a.segment, a.column, a.row_offset).cmp(&(b.segment, b.column, b.row_offset))
```

adr1anh commented on codegen/ace/src/comparison.rs:
```diff
@@ -0,0 +1,64 @@
+    }
+}
+
+fn value_order(a: &Value) -> usize {
```
This and the methods below look like an implementation of [`PartialOrd`](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html) which is like `Ord` but instead of `cmp(..) -> Ordering` we have `partial_cmp(..) -> Option<Ordering>`. Returning an `Option<usize>` here instead of `unreachable` would shift the responsibility to the caller to `unwrap` if they can ensure the ordering is always valid.

adr1anh commented on codegen/ace/src/comparison.rs:
```diff
@@ -0,0 +1,64 @@
+    }
+}
+
+fn compare_value(a: &Value, b: &Value) -> Ordering {
```
Same here, returning `Option<Ordering>`

adr1anh commented on codegen/ace/src/comparison.rs:
```diff
@@ -0,0 +1,64 @@
+
+/// Compare functions used to sort the operations as needed by the ace chiplet.
+pub fn compare_operation(a: &Operation, b: &Operation) -> Ordering {
+    let cmp = value_operation(a).cmp(&value_operation(b));
```
Since `Operation` already derives `Ord`, you could just set `cmp = a.cmp(b)`, since the derived implementation guarantees lexicographical comparison of struct fields.

paracetamolo replied:
I don't want lexicographic ordering, I want values to be first and the rest untouched. The sorting algorithm is stable so this should work.

adr1anh commented on air/src/graph/mod.rs:
```diff
@@ -23,6 +23,18 @@ impl core::ops::Add<usize> for &NodeIndex {
     }
 }
 
+impl From<usize> for NodeIndex {
```
There is a blanket implementation `Into<U> for T` when we implement `From<T> for U` [source](https://doc.rust-lang.org/std/convert/trait.From.html#generic-implementations). This mean you can write
```rust
let n = NodeIndex::default();
let u: usize = n.into().
```

adr1anh commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,507 @@
+/// Remap the children of an operation. Typically when migrating an operation
+/// between data structures.
+pub fn remap_op(op: Operation, mapping: &BTreeMap<NodeIndex, NodeIndex>) -> Operation {
+    use Operation::*;
```
Importing all enum variants is generally discourages since it may lead to incorrect matches when the enum variants change. In this case it would be okay, but if the variants did not have fields inside them, a removed field would be matched as the "catch-all" case.

paracetamolo commented:
Done:
- updated formula as in point 7 of
https://github.com/0xPolygonMiden/air-script/issues/366#issuecomment-2751848648
- fixed serialization format
- tested circuits with periodic columns

Todo:
- documenting some functions
- there is something fishy with public inputs, need to test it better

adr1anh commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,507 @@
+    pub fn new() -> Self {
+        let mut cache: BTreeMap<Operation, NodeIndex> = BTreeMap::new();
+        let mut circuit = Circuit::new();
+        let constants: Vec<NodeIndex> = [0, 1]
+            .iter()
+            .map(|c| {
+                let op = Operation::Value(Value::Constant(*c as u64));
+                let out = circuit.append(op);
+                cache.insert(op, out);
+                out
+            })
+            .collect();
+        Self {
+            cache,
+            circuit,
+            n_inputs: 0,
+            zero: constants[0],
+            one: constants[1],
+            alpha: None,
+            next_alpha: constants[1],
+        }
```
Arrays have a `map` method that returns an array of the same size. It can be destructured in-place. This avoids an extra `Vec` allocation
```suggestion
        let [zero, one] = [0, 1].map(|c| {
            let op = Operation::Value(Value::Constant(c as u64));
            let out = circuit.append(op);
            cache.insert(op, out);
            out
        });
        Self {
            cache,
            circuit,
            n_inputs: 0,
            zero,
            one,
            alpha: None,
            next_alpha: one,
        }
```

paracetamolo replied:
Very nice. Done.

adr1anh commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,507 @@
+
+    /// Append an operation to the circuit after applying some optimizations
+    /// from the algebraic laws of addition, multiplication and subtraction.
+    pub fn append(&mut self, op: Operation) -> NodeIndex {
```
Again, `push` would be match `Vec`

paracetamolo replied:
done

adr1anh commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,507 @@
+    /// `\sum_i vec[i]`
+    pub fn sum(&mut self, vec: &[NodeIndex]) -> NodeIndex {
+        assert!(!vec.is_empty());
+        vec.iter().fold(self.zero, |acc, r| self.add(acc, *r))
```
I think adding zero is optimized, but `reduce` ensures that we don't perform a trivial addition.

```suggestion
        vec.iter().reduce(|acc, r| self.add(acc, *r)).unwrap_op(self.zero)
```

paracetamolo replied:
Nice to learn about reduce. I tried but it gets messy with the references. I would leave it a fold for now.

adr1anh commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,507 @@
+    /// `\prod_i vec[i]`
+    pub fn prod(&mut self, vec: &[NodeIndex]) -> NodeIndex {
+        assert!(!vec.is_empty());
+        vec.iter().fold(self.one, |acc, r| self.mul(acc, *r))
```
Same as above
```suggestion
        vec.iter().reduce(|acc, r| self.mul(acc, *r)).unwrap_op(self.one)
```

adr1anh commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,507 @@
+    /// using Horner's method.
+    /// `\sum_i coeffs[i] * x^i`
+    pub fn horners_method(&mut self, point: NodeIndex, coeffs: &[NodeIndex]) -> NodeIndex {
+        coeffs.iter().rev().fold(self.zero, |acc, coeff| {
```
Same as above for `reduce`

paracetamolo replied:
Same as above

adr1anh reviewed (commented):
Not a full review of the functionality, but left some code-related comments as I tried getting familiar with the circuit.

paracetamolo commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,509 @@
+        (self.0.len() - 1).into()
+    }
+
+    fn iter(&self) -> core::slice::Iter<Operation> {
```
Right. I removed this code now, but good note for the future.

paracetamolo commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,509 @@
+        self.0.is_empty()
+    }
+
+    fn append(&mut self, op: Operation) -> NodeIndex {
```
done

paracetamolo commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,509 @@
+        self.0.iter()
+    }
+
+    fn get_inputs(&self) -> Vec<NodeIndex> {
```
Right. I removed this code now, but good note for the future.

paracetamolo commented:
Update:
- I did a large refactor of input handling. Before it was possible to define inputs anywhere during the circuit definition and they would be ordered later. Right now all inputs must be defined at circuit creation, which simplifies some things.
- Defined the arrangement of all inputs of the Airscript and ACE circuits.
- Improved testing a lot, typically to correctly test public inputs, random values, and next-row values.

paracetamolo commented:
Note on future work.
This PR uses the `Operation` type because it was rich in information and we were not entirely sure of the design.
Now that the design is clearer, a new much simpler type can be defined with simply `Add, Mul, Sub` so that most of the unreachable patter matchings can be removed and also `NodeIndex` which causes a bit of overhead.
This is left as future work to be done at the same time as VM trace generation, which will share code.

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,342 @@
+    let lhs = {
+        let mut els = vec![];
+        // int * z_g^2 * z_1
+        if !integrity_roots.is_empty() {
+            let integrity_roots = integrity_roots.iter().map(|r| mapping[r]).collect();
+            let lc = cb.linear_combination_alpha(integrity_roots);
+            let res = cb.prod(&[lc, z_minus_g, z_minus_g, z_minus_one]);
+            els.push(res)
+        };
+        // bf * zn_1 * z_g
+        if !boundary_first_roots.is_empty() {
+            let boundary_first_roots = boundary_first_roots.iter().map(|r| mapping[r]).collect();
+            let lc = cb.linear_combination_alpha(boundary_first_roots);
+            let res = cb.prod(&[lc, z_n_minus_one, z_minus_g]);
+            els.push(res)
+        };
+        // bl * zn_1 * z_1
+        if !boundary_last_roots.is_empty() {
+            let boundary_last_roots = boundary_last_roots.iter().map(|r| mapping[r]).collect();
+            let lc = cb.linear_combination_alpha(boundary_last_roots);
+            let res = cb.prod(&[lc, z_n_minus_one, z_minus_one]);
+            els.push(res)
+        };
```
But isn't `CircuitBuilder::next_alpha()` there for exactly this purpose? or maybe there is something else?

paracetamolo replied:
Unless I made I mistake somewhere, `linear_combination_alpha` keeps the successive `alpha` stored in the `CircuitBuilder` state. So the three calls use always different alphas. I was undecided if to move the alpha as an argument and to pass them explicitly to make the code clearer but at the cost of making a mistake.
I added a test to check that two calls to `linear_combinarion_alpha` using the same coeffs and input lead to different results.

paracetamolo commented:
Rebased branch on `next`, after busses. Just need to squash the history.

Al-Kindi-0 commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,487 @@
+    }
+
+    // Append an operation to the circuit. A new operation is added only if
+    // syntactically different from any operations already added to the circuit
```
nit: "... if it is syntactically different from any operation already added ..."

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,345 @@
+/// - the aux segment of trace inputs of the AirScript e.g. `trace_columns { aux: [f] }`
+/// - the next row main segment of trace inputs of the AirScript e.g. `a' b'`
+/// - the next row aux segment of trace inputs of the AirScript e.g. `f'`
+/// - qi(z) with i in [0,7)
```
the range should be [0, 8)

Al-Kindi-0 replied:
@paracetamolo , could we fix this?

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,345 @@
+/// - z
+/// - z^n
+/// - g^(-1)
+/// - z^min_num_cycles
```
nit: maybe the formula computing min_num_cycles should be added

Al-Kindi-0 replied:
@paracetamolo , could we fix this?

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,345 @@
+/// - qi(z) with i in [0,7)
+/// - alpha
+/// - z
+/// - z^n
```
nit: maybe mention that n is the execution trace length

Al-Kindi-0 replied:
@paracetamolo , could we fix this?

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,345 @@
+        UniqueOperationIterator::new(ir.constraint_graph(), all_roots)
+            .partition(|(_, op)| matches!(op, Operation::Value(Value::PeriodicColumn(_))));
+
+    // A circuit builder is intantiated with the inputs of the circuits plus the 13 needed by the ACE chiplet
```
typo: instantiated

Al-Kindi-0 replied:
@paracetamolo , could we fix this?

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,345 @@
+
+    // A circuit builder is intantiated with the inputs of the circuits plus the 13 needed by the ACE chiplet
+    let n_inputs = n_random(ir) + n_public_inputs(ir) + n_trace_accesses(ir);
+    let (mut cb, inputs) = CircuitBuilder::new(n_inputs + 13);
```
Maybe we should use a constant and define how it is computed?

Al-Kindi-0 replied:
@paracetamolo , could we fix this?

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,345 @@
+    // A circuit builder is intantiated with the inputs of the circuits plus the 13 needed by the ACE chiplet
+    let n_inputs = n_random(ir) + n_public_inputs(ir) + n_trace_accesses(ir);
+    let (mut cb, inputs) = CircuitBuilder::new(n_inputs + 13);
+    let qs: Vec<NodeIndex> = inputs[n_inputs..n_inputs + 8].to_vec();
+    let alpha = inputs[n_inputs + 8];
+    cb.alpha = Some(alpha);
+    let z_point = inputs[n_inputs + 9];
+    let z_n = inputs[n_inputs + 10];
+    let inverse_generator = inputs[n_inputs + 11];
+    let z_min_num_cycles = inputs[n_inputs + 12];
```
Same comment regarding constants here

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,345 @@
+    // A mapping is necessary to convert the indexes of the Air to the new indexes of the circuit
+    let mut mapping: BTreeMap<NodeIndex, NodeIndex> = BTreeMap::new();
+
+    periodic_columns(ir, &mut cb, &mut mapping, periodic_ops, z_min_num_cycles);
```
nit: could we add a comment here?

Al-Kindi-0 replied:
@paracetamolo , could we fix this?

Al-Kindi-0 replied:
nit: could we add a comment here ?

paracetamolo replied:
done

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,345 @@
+
+    // Inputs are filtered and their indexes are converted. They are not added
+    // to the circuit because, unlike in the algebraicgraph where inputs are
+    // nodes in the graph, in the ACE circuit inputs are inplicit. They are
```
implicit?

Al-Kindi-0 replied:
@paracetamolo , could we fix this?

paracetamolo replied:
done

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,345 @@
+    // nodes in the graph, in the ACE circuit inputs are inplicit. They are
+    // simply the first indexes.
+    // Their converted indexes however need to be added to the mapping so that
+    // any nodes using an index can be remapped correctly.
```
nit: " ... any nodes using such an index ..."

paracetamolo replied:
done

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,345 @@
+        mapping.insert(out, new_out);
+    }
+
+    // The original Airscript is not embedded inside the ACE circuit.
+    // We start adding new nodes to join the Airscript roots in the formula
+    // described above.
```
Could you reformulate/explain more here?

paracetamolo replied:
It's a very confusing typo, it was supposed to be "is noW embedded". I rephrased it.

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,345 @@
+            })
+            .collect();
+
+        // The polynomials are stored in the circuit as constants
```
nit: The coefficients of the interpolated polynomials

paracetamolo replied:
done

Al-Kindi-0 commented on codegen/ace/src/serialization.rs:
```diff
@@ -0,0 +1,178 @@
+///
+/// In the case of a gate, the 63 bits available in a field element are used in
+/// the following order, from least significant:
+/// - 20 bits for right index
+/// - 20 bits for left index
```
I think we now use 30 bits each, @adr1anh ?

paracetamolo replied:
It's not the comment, it's the code that was outdated, fixed it now.

Al-Kindi-0 commented on codegen/ace/src/serialization.rs:
```diff
@@ -0,0 +1,178 @@
+/// 2, // two gates
+/// // end header
+/// 14, // value of the constant, which has index 1, given that there is one input at index 0
+/// (2 << 40) + (0 << 20) + (1 << 0), // addition gate opcode with left intput 0 and right input 1. Output index 2.
+/// (1 << 40) + (2 << 20) + (0 << 0), // multiplication gate opcode with left intput 2 and right input 0. Output index 3.
```
same here

paracetamolo replied:
done

Al-Kindi-0 reviewed (changes requested):
Looks great!

There are a few optimizations improvement that we can do, but I thin we should just try to get this merged asap. This way we can progress the work on the VM side while at the same time open PRs here and elsewhere in order to clean up and improve the codegen pipeline.

paracetamolo commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,487 @@
+        }
+    }
+
+    pub fn to_dot(&self) -> Result<String, std::fmt::Error> {
+        operations_to_dot(&self.ops)
+    }
```
done

paracetamolo commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,487 @@
+        operations_to_dot(&self.ops)
+    }
+
+    /// Serialize the circuit to a vector of field elements, as expected by the
```
done

paracetamolo commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,487 @@
+    // nodes already processed, meaning their children are visited
+    processed: BTreeSet<NodeIndex>,
+    // nodes to be processed
+    process: Vec<NodeIndex>,
```
done

paracetamolo commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,487 @@
+    }
+}
+
+/// Remap the children of an operation. Typically when migrating an operation
+/// between data structures. Any index withtin the inputs is left untouched.
```
done

paracetamolo commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,487 @@
+    // A cache of nodes already inserted in the circuit, used to avoid duplicates
+    cache: BTreeMap<Operation, NodeIndex>,
+    // Constants zero and one are inserted first so that they can be used as
+    // identity and absorbing element in addition and multiplication.
```
done

paracetamolo commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,345 @@
+    let z_minus_one = cb.sub(z_point, cb.one);
+    let z_n_minus_one = cb.sub(z_n, cb.one);
+
+    let lhs = {
```
I created a separate struct `linear_combination_alpha`

paracetamolo commented:
Update:
- addressed all the comments
- moved public inputs before random
- renamed all regression files with the more informative script name rather than a number e.g. 1.dot -> ComplexBoundary.dot

adr1anh commented:
Looks good to me!

Al-Kindi-0 commented on codegen/ace/src/circuit.rs:
```diff
@@ -0,0 +1,457 @@
+use std::collections::{btree_map::BTreeMap, btree_set::BTreeSet};
+
+/// A circuit that can be consumed by the ACE chiplet.
+/// The only way to build a circuit is through the CircuitBuilder, it can then obtained
```
"can then be obtained" ?

paracetamolo replied:
The only way to obtain an object of type Circuit is from the normalize function of CircuitBuilder. I'm trying to explain the flow of usage. I can rephrase if it is confusing.

Al-Kindi-0 replied:
the comment was only on a typo in "obtained" which is now fixed, thank you

Al-Kindi-0 commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,406 @@
+        None
+    }
+}
+
+// Helper functions to convert the Air inputs to ACE inputs.
```
Should we a section separator as we done in other repositories?

paracetamolo replied:
done

Al-Kindi-0 reviewed (changes requested):
Left some new and old comments, once these are addressed we can merge in say a `ace-codegen` branch.

adr1anh commented on codegen/ace/src/codegen.rs:
```diff
@@ -0,0 +1,406 @@
+/// The circuit gets as input `z^min_num_cycles`.
+/// For a given cycle_len let `k = max_cycle_len / cycle_len`, then `z^l`, where
+/// `l := trace_len / cycle_len`, is nothing but `(z^min_num_cycles)^k`.
+fn periodic_columns(
```
It might be clearer to factor out a method for computing the node of a single periodic column, wrapping it in a loop that computes all cycles. This can be done in a future refactor PR.

```rust
        // Maximum length of all periodic columns in the air. 
        let max_cycle_len = self
            .air
            .periodic_columns
            .values()
            .map(|pi| pi.values.len())
            .max()
            .unwrap();

        // The power of `z` for the longest column. Let `l` such that `z_min_cycle = z^l`, 
        // such that `l = trace_len / max_cycle_len`
        let z_min_cycle = Node::Input(self.layout.air_input_index(Z_MIN_CYCLE));
        
        let cycle_len = periodic_column.values.len();
        assert!(max_cycle_len.is_power_of_two());
        assert!(cycle_len.is_power_of_two());
        let log_k = (max_cycle_len / cycle_len).ilog2();
        let z_pow_l = (0..log_k).fold(z_min_cycle, |acc, _| self.square(acc));

        let mut column: Vec<_> = periodic_column
            .values
            .iter()
            .copied()
            .map(Felt::new)
            .collect();
        let inv_twiddles = winter_math::fft::get_inv_twiddles::<Felt>(column.len());
        winter_math::fft::interpolate_poly(&mut column, &inv_twiddles);

        let poly: Vec<_> = column
            .into_iter()
            .map(|c| self.get_constant_node(u64::from(c)))
            .collect();

        let result = self.horner(z_pow_l, poly.into_iter().rev());
```

Al-Kindi-0 replied:
Maybe we should create an issue on this?

paracetamolo reviewed (commented):
Fixed some old comments that I had missed.
